src/analyses/goto_check.cpp

```diff
@@ -1439,24 +1439,26 @@ void goto_checkt::check_rec(const exprt &expr, guardt &guard, bool address)
   }
   else if(expr.id()==ID_and || expr.id()==ID_or)
   {
-    if(!expr.is_boolean())
+    if(ns.follow(expr.type()).id() != ID_bool)
       throw "`"+expr.id_string()+"' must be Boolean, but got "+
             expr.pretty();
 
     guardt old_guard=guard;
 
     for(const auto &op : expr.operands())
     {
-      if(!op.is_boolean())
+      auto op_followed = op;
+      op_followed.type() = ns.follow(op_followed.type());
+      if(op_followed.type().id() != ID_bool)
         throw "`"+expr.id_string()+"' takes Boolean operands only, but got "+
-              op.pretty();
+              op_followed.pretty();
 
-      check_rec(op, guard, false);
+      check_rec(op_followed, guard, false);
```

Comment:
Why keep the resolved type rather than basing subexpressions on `op`? If you're trying to exclude symbol types that resolve to primitives because cbmc is fragile in the face of those then you should probably do that in a specific pass upon loading the symbol table, rather than doing it ad-hoc like this. Otherwise if you're happy with the symbol types but `is_boolean` isn't clever enough then only use the resolved type where you need it.